satellite/metainfo/loop.go

```diff
@@ -0,0 +1,235 @@
+	join     chan *observerContext
+	done     chan struct{}
+	cancel   func()
+}
+
+// NewLoop creates a new metainfo loop service.
+func NewLoop(config LoopConfig, metainfo *Service) *Loop {
+	return &Loop{
+		metainfo: metainfo,
+		config:   config,
+		join:     make(chan *observerContext),
+		done:     make(chan struct{}),
+	}
+}
+
+// Join will join the looper for one full cycle until completion and then returns.
+// On ctx cancel the observer will return without completely finishing.
+// Only on full complete iteration it will return nil.
+func (loop *Loop) Join(ctx context.Context, observer Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obsContext := &observerContext{
+		Observer: observer,
+		ctx:      ctx,
+		done:     make(chan error),
+	}
+
+	select {
+	case loop.join <- obsContext:
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-loop.done:
+		return LoopClosedError
+	}
+
+	return obsContext.Wait()
+}
+
+// Run starts the looping service.
+func (loop *Loop) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	ctx, cancel := context.WithCancel(ctx)
+	loop.cancel = cancel
+	defer close(loop.done)
+
+	for {
+		err := loop.runOnce(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// runOnce goes through metainfo one time and sends information to observers
+func (loop *Loop) runOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var observers []*observerContext
+	defer func() {
+		for _, observer := range observers {
+			observer.Finish()
+		}
+	}()
+
+	// wait for the first observer, or exit because context is canceled
+	select {
+	case observer := <-loop.join:
+		observers = append(observers, observer)
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	// after the first observer is found, set timer for CoalesceDuration and add any observers that try to join before the timer is up
+	timer := time.NewTimer(loop.config.CoalesceDuration)
+waitformore:
+	for {
+		select {
+		case observer := <-loop.join:
+			observers = append(observers, observer)
+		case <-timer.C:
+			break waitformore
+		case <-ctx.Done():
+			for _, observer := range observers {
+				observer.HandleError(ctx.Err())
+			}
+			observers = nil
+			return ctx.Err()
+		}
+	}
+
+	err = loop.metainfo.Iterate(ctx, "", "", true, false,
+		func(ctx context.Context, it storage.Iterator) error {
+			var item storage.ListItem
+
+			// iterate over every segment in metainfo
+			for it.Next(ctx, &item) {
+				pointer := &pb.Pointer{}
+
+				err = proto.Unmarshal(item.Value, pointer)
+				if err != nil {
+					// TODO: figure out what to do
+					// return LoopError.New("error unmarshalling pointer %s", err)
+					continue
+				}
+
+				path := item.Key.String()
+				pathElements := storj.SplitPath(path)
+				isLastSeg := len(pathElements) >= 2 && pathElements[1] == "l"
+
+				nextObservers := observers[:0]
+
+				for _, observer := range observers {
+					// TODO: move single observer handling into a separate func
```

Comment:
I am removing because since handling a single observer involves loop logic and it would probably make things more confusing if we had another function for this